### PR TITLE
Feat(mlx): Add eval mean_token_accuracy metric to MLXTrainer (baseline path)

### PR DESCRIPTION
**Pull request by BardiaKoopah**
## Summary
Adds `eval_mean_token_accuracy` to MLXTrainer eval, matching TRL's
SFTTrainer metric so MLX runs are comparable to CUDA/Colab runs.

## Scope
- Computed only on the baseline (non-CCE) loss path, where full logits
  exist. Under CCE the loss returns (loss, lse) with no logits, so the
  metric is omitted — mirroring TRL's own behavior with its fused Liger
  kernel (it skips logit materialization unless explicitly needed).
- Eval-only. Training-loop and value_and_grad paths are untouched: the
  loss fns gain an opt-in `return_correct=False` kwarg, so the
  (loss, ntoks) contract the grad wrapper depends on is unchanged.

## Implementation
- Four loss fns (utils.py) take `return_correct=False`; baseline paths
  return (loss, ntoks, correct) when set, CCE paths return (..., None).
- `_evaluate_batch_totals` gains `want_accuracy`; `_evaluate` threads
  correct-token totals and emits `eval_mean_token_accuracy`.

## Testing
5-step gemma-3-4b-it VLM LoRA run, baseline + CCE:
- baseline: eval_mean_token_accuracy ≈ 0.456 reported
- CCE: key correctly absent; eval_loss identical across both (4.1048)

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/trainer.py [hunk 1]
```diff
@@ -1549,17 +1569,30 @@ def _evaluate(self, eval_batches, loss_fn, is_vlm=False):
...
             )
             all_losses = self._distributed_all_sum(all_losses, stream=mx.cpu)
             ntokens = self._distributed_all_sum(ntokens, stream=mx.cpu)
+            # DDP: reduce the token-accuracy numerator across ranks before
+            # dividing. ntokens above is already the global token denominator,
+            # so both correct_total and its denominator are rank-summed here.
+            if correct_total is not None:
+                correct_total = self._distributed_all_sum(
+                    correct_total, stream=mx.cpu,
+                )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: Fixed in 0027315: the accuracy numerator all-reduce is now unconditional across ranks. It reduces a rank-wide any-numerator consensus flag first, then every rank all-sums a zero-or-real numerator, so a rank that produced no numerator (an external stop before its first eval batch, or a loss backend that returns no argmax) no longer skips the collective while peers call it and deadlocks.

File: unsloth_zoo/mlx/trainer.py [hunk 2]
```diff
@@ -1496,12 +1498,29 @@ def _evaluate_batch_totals(self, eval_batches, loss_fn, is_vlm=False):
...
+                            )
+                        else:
+                            loss, ntoks = loss_fn(self.model, batch, lengths, labels)
+                            c = None
                     all_losses += loss * ntoks
                     ntokens += ntoks
+                    if c is not None:
+                        correct_total = (
+                            c if correct_total is None else correct_total + c
+                        )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py:3590]: Fixed in 0027315: correct_total is now materialized alongside the loss and token totals each batch, so the per-batch argmax(logits) graph no longer stays live until the final metric and eval memory no longer grows with the whole eval set.

File: unsloth_zoo/mlx/trainer.py [hunk 3]
```diff
@@ -1526,12 +1545,13 @@ def _evaluate(self, eval_batches, loss_fn, is_vlm=False):
...
+        correct_total = None
         if isinstance(eval_batches, dict):
             all_losses = mx.array(0.0)
             ntokens = mx.array(0)
             for split_name, split_batches in eval_batches.items():
-                split_losses, split_tokens = self._evaluate_batch_totals(
-                    split_batches, loss_fn, is_vlm=is_vlm,
+                split_losses, split_tokens, _ = self._evaluate_batch_totals(
+                    split_batches, loss_fn, is_vlm=is_vlm, want_accuracy=False,
                 )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: Fixed in 0027315: the dict eval path now enables accuracy as well, accumulating the numerator across splits and applying the same rank-consistent reduce, so eval_mean_token_accuracy is reported for dict datasets and metric_for_best_model=eval_mean_token_accuracy works there too.

File: unsloth_zoo/mlx/trainer.py [hunk 4]
```diff
@@ -1549,17 +1587,44 @@ def _evaluate(self, eval_batches, loss_fn, is_vlm=False):
...
         avg_loss = (all_losses / ntokens).item() if ntokens.item() > 0 else 0.0
         perplexity = math.exp(min(avg_loss, 100))
         metrics["eval_loss"] = avg_loss
         metrics["eval_perplexity"] = perplexity
+        if correct_total is not None:
+            token_denominator = _safe_token_denominator(ntokens)
+            mx.eval(correct_total, token_denominator)
+            metrics["eval_mean_token_accuracy"] = (
+                correct_total / token_denominator
+            ).item()
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: Good catch. Fixed in f316fb1: greater_is_better now defaults to None and is inferred from the metric name (loss/perplexity are lower-is-better, accuracy and others higher-is-better), matching HF TrainingArguments, so selecting eval_mean_token_accuracy for metric_for_best_model no longer minimizes it. An explicit greater_is_better still wins.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
## Validation on real Apple Silicon (Metal)

Ran this branch end to end on a macOS Apple Silicon runner with real MLX on a real Metal GPU (no torch shim) via staging CI (head `78a88c8`). Stack: mlx 0.32.0, mlx-metal 0.32.0, mlx-lm 0.31.3, transformers 5.13.0.

### Real-MLX feature validation

| Check | Result |
|---|---|
| real MLX / Metal available | PASS |
| import unsloth_zoo.mlx | PASS |
| pr820_token_accuracy (API) | PASS |
| pr820_eval_accuracy_training (real training) | PASS |
| shared LoRA SFT smoke | PASS |
| Total | 5 pass, 8 skip, 0 fail |

### Token-accuracy training smoke on Metal (LoRA SFT, 6 steps)

| Step | Loss | Grad | Tok/s | Peak mem |
|---|---|---|---|---|
| 1 | 3.5565 | 1.4059 | 37 | 0.25 GB |
| 2 | 3.8368 | 1.4983 | 78 | 0.26 GB |
| 3 | 3.8277 | 1.4749 | 80 | 0.26 GB |
| 4 | 3.7257 | 1.7337 | 183 | 0.26 GB |
| 5 | 3.1310 | 1.7143 | 91 | 0.27 GB |
| 6 | 3.0060 | 1.6977 | 250 | 0.27 GB |

Avg loss 3.5139, 235 tokens, wall 4.1s. The accuracy path runs with the token-accuracy metric enabled and matches the shared smoke curve to within the added masked-accuracy bookkeeping.

**Comment by danielhanchen** [comment]
## Merged with main and re-validated on Metal

This branch conflicted with main after #888 (guard MLX distributed eval against zero-token batches) landed, since both touch the MLX eval loss path. Merged main in and resolved:

- `mlx/utils.py` `make_baseline_loss_fn` (`labels is None`): kept main's raw-`ntoks` denominator for byte-for-byte `mlx_lm` `default_loss` parity, with the token-accuracy `return_correct` branch on top. `_safe_token_denominator` stays on the labels-aware / CCE paths as main left it.
- `mlx/trainer.py` `_evaluate`: kept the `want_accuracy` branch (which produces the per-batch correct count the accuracy total consumes) and adopted main's NaN-safe `all_losses += mx.where(ntoks > 0, loss * ntoks, 0.0)`, so zero-token padding batches cannot poison the distributed all-sum. The guard now also covers the VLM eval path.

Re-validated the merged head (`db2e4bb8`) on a macOS Apple Silicon runner with real MLX / Metal (mlx 0.32.0, mlx-metal 0.32.0, mlx-lm 0.31.3, transformers 5.13.0):

| Check | Result |
|---|---|
| pr820_token_accuracy | PASS |
| pr820_eval_accuracy_training | PASS |
| shared LoRA SFT smoke | PASS |
| Total | 5 pass, 8 skip, 0 fail |

Token-accuracy eval training smoke on Metal (6 steps): loss 3.5565 to 3.0060, avg 3.5139, 235 tokens, peak 0.27 GB, wall 3.4s. Now MERGEABLE with no conflicts. 93 `tests/test_mlx_trainer_internals.py` pass off-Mac.

**Comment by danielhanchen** [comment]
@codex review

**Comment by BardiaKoopah** [comment]
Small note on the `greater_is_better` inference from `f316fb1`, since it's easy to miss: the rule here is `not (endswith("loss") or endswith("perplexity"))`, which is deliberately one case stricter than HF's `TrainingArguments`. HF uses `not endswith("loss")`, so upstream would infer `eval_perplexity` as higher-is-better and `load_best_model_at_end` would keep the highest-perplexity checkpoint. Since `_evaluate` emits `eval_perplexity` on every eval, that case is reachable here, so the extra suffix is intentional and pinned by `test_greater_is_better_inferred_from_metric_name`. Flagging it only so a future "align with HF" refactor doesn't drop it as redundant.

**Comment by BardiaKoopah** [comment]
Status: done from my side — this one is converged and good to merge.

All four Codex P2s from the 07-08 rounds are resolved (`0027315`, `f316fb1`): dict-eval accuracy, per-batch numerator materialization, the rank-consistent all-reduce, and `greater_is_better` inference from the metric name. Codex has since returned "no major issues" twice, most recently on `db2e4bb8` after main was merged in to resolve the #888 conflict. Metal re-validation after that merge passed (token-accuracy API, eval-accuracy training, shared LoRA SFT smoke — 5 pass, 8 skip, 0 fail), and the off-Mac trainer-internals suite is green.

CI is green and the branch is mergeable. Happy to rebase if anything has landed since.

**Comment by BardiaKoopah** [comment]
@danielhanchen 
Done from my side. I went through the Codex findings. The "greater_is_better" one you fixed in f316fb1 is correct and Codex is clean. Added a note too for why the perplexity handling is intentionally one case stricter than HF so it doesn't get "simplified" away later. Good to merge whenever.

**Comment by BardiaKoopah** [comment]
@codex review

**Comment by BardiaKoopah** [comment]
@Daniel ready to merge

**Comment by BardiaKoopah** [comment]
@codex review
